schemachanger: implement `ALTER TABLE .. VALIDATE CONSTRAINT ..`

Xiang-Gu opened the pull request:
The main idea is to drop the unvalidated element and add a vanilla counterpart in the builder state.

Commit 2 involves some changes to allow both legacy and declarative schema changer handle
this stmt correctly: `ALTER TABLE .. ADD/DROP CONSTRAINT .. [NOT VALID]; VALIDATE CONSTRAINT ..;`
so this PR warrants a release note.

Epic: None
Release note (bug fix): Allow `ALTER TABLE .. ADD/DROP CONSTRAINT .. [NOT VALID]; VALIDATE CONSTRAINT ..;` to behave consistently with Postgres. Previously, the VALIDATE CONSTRAINT part will fail and cause the whole statement to fail.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96648)
<!-- Reviewable:end -->

Xiang-Gu committed:
schemachanger: Implement `ALTER TABLE .. VALIDATE CONSTRAINT`

The main idea is to drop the unvalidated element and add a vanilla
counterpart in the builder state.

Xiang-Gu commented:
This is ready for a look!

ajwerner reviewed (commented):
> Reviewed 15 of 15 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96648)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCDJu0QJW87OnMjp3:-NNhCDJu0QJW87OnMjp4:bqnzfjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L44)):*
> ```Go
> 	}
> 	if retrieveUniqueWithoutIndexConstraintElem(b, tbl.TableID, constraintID) != nil {
> 		return
> ```

Consider factoring out a helper here called `getConstraintElementByID`. Also, is there an issue if the constraint had been dropped? Like imagine you dropped the constraint and then validated it? Would you find it?



_Code quote:_
```Go
	if retrieveCheckConstraintElem(b, tbl.TableID, constraintID) != nil {
		return
	}
	if retrieveUniqueWithoutIndexConstraintElem(b, tbl.TableID, constraintID) != nil {
		return
	}
	if retrieveForeignKeyConstraintElem(b, tbl.TableID, constraintID) != nil {
		return
	}
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCctrE_101Ia17f91:-NNhCctrE_101Ia17f92:b9mi7gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L57)):*
> ```Go
> 		uwiNotValidElem:    retrieveUniqueWithoutIndexConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
> 		fkNotValidElem:     retrieveForeignKeyConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
> 	}
> ```

nit: any reason to not just construct the spec inline? I feel like it's slightly more pleasant to not introduce the intermediate variable. 



_Code quote:_
```Go
	spec := validateConstraintSpec{
		constraintNameElem: constraintNameElem,
		ckNotValidElem:     retrieveCheckConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
		uwiNotValidElem:    retrieveUniqueWithoutIndexConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
		fkNotValidElem:     retrieveForeignKeyConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
	}
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhDFfb-ZCf8EzJJnxK:-NNhDFfb-ZCf8EzJJnxL:bmsqir2) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L61)):*
> ```Go
> }
> 
> func retrieveCheckConstraintElem(
> ```

These smell, but I don't have anything better for you right now. Just a passing thought. 


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96648)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @rytaft)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCDJu0QJW87OnMjp3:-NNhSlgz5DrAiY9y3ixc:bjclgkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L44)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Consider factoring out a helper here called `getConstraintElementByID`. Also, is there an issue if the constraint had been dropped? Like imagine you dropped the constraint and then validated it? Would you find it?
</blockquote></details>

1. Done. I instead created a helper where we determine whether or not we should validate the constraint.

2.
 - if the constraint had just been dropped: the constraint name will be set to a placeholder and we will return an error from `resolveConstraint`, consistent with what the legacy schema changer does.
 - if the constraint had just been added: the logic will determine that this is something we should validate and thus it can proceed.



_Code quote:_
```Go
if retrieveUniqueWithoutIndexConstraintElem(b, tbl.TableID, constraintID) != nil {
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCctrE_101Ia17f91:-NNhTpOFEJ6rxbEM93JQ:b-qdxsyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L57)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: any reason to not just construct the spec inline? I feel like it's slightly more pleasant to not introduce the intermediate variable. 
</blockquote></details>

nothing particularly special: I just feel like the resulting function signature doesn't look pretty
```validateConstraint(b, tableID, constraintID, constraintNameElem)```


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96648)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCDJu0QJW87OnMjp3:-NNhXkP246_mNedGoklX:b-vhfd9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L44)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

1. Done. I instead created a helper where we determine whether or not we should validate the constraint.

2.
 - if the constraint had just been dropped: the constraint name will be set to a placeholder and we will return an error from `resolveConstraint`, consistent with what the legacy schema changer does.
 - if the constraint had just been added: the logic will determine that this is something we should validate and thus it can proceed.
</blockquote></details>

When you say `the constraint name will be set to a placeholder`, where does that happen? 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCctrE_101Ia17f91:-NNhXYsdB2WdiHbDOAaM:b-r2rroz) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L57)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

nothing particularly special: I just feel like the resulting function signature doesn't look pretty
```validateConstraint(b, tableID, constraintID, constraintNameElem)```
</blockquote></details>

I was just suggesting:

```go
validateConstraint(b, tbl.TableID, validateConstraintSpec{
	constraintNameElem: constraintNameElem,
	ckNotValidElem:     retrieveCheckConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
	uwiNotValidElem:    retrieveUniqueWithoutIndexConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
	fkNotValidElem:     retrieveForeignKeyConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
})
```


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96648)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @rytaft)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCDJu0QJW87OnMjp3:-NNhf8UC-eMJTSemyqWJ:b-gmxki1) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L44)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

When you say `the constraint name will be set to a placeholder`, where does that happen? 
</blockquote></details>

In the statement phase of that `DROP CONSTRAINT`, that not-valid constraint will be dropped as well as its name.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCctrE_101Ia17f91:-NNhfPz64YLevpOAUl0D:b-468704) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L57)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I was just suggesting:

```go
validateConstraint(b, tbl.TableID, validateConstraintSpec{
	constraintNameElem: constraintNameElem,
	ckNotValidElem:     retrieveCheckConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
	uwiNotValidElem:    retrieveUniqueWithoutIndexConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
	fkNotValidElem:     retrieveForeignKeyConstraintUnvalidatedElem(b, tbl.TableID, constraintID),
})
```
</blockquote></details>

ah I see . done


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96648)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCDJu0QJW87OnMjp3:-NNi88J0AwQisgkq65rC:b-em7pd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L44)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

In the statement phase of that `DROP CONSTRAINT`, that not-valid constraint will be dropped as well as its name.
</blockquote></details>

So what happens if you do:

```sql
CREATE TABLE t (i INT PRIMARY KEY);
ALTER TABLE t ADD CONSTRAINT c CHECK (i > 0) NOT VALID;
```
and
```sql 
ALTER TABLE t DROP CONSTRAINT c, VALIDATE CONSTRAINT c;
```
Or:
```sql
ALTER TABLE t DROP CONSTRAINT c, ADD CONSTRAINT c CHECK (i > 1) NOT VALID, VALIDATE CONSTRAINT c;
```
Does the first case error and the second case validate the newly added check constraint?


<!-- Sent from Reviewable.io -->

Xiang-Gu committed:
sqlerrors: Refactor out a helper function for creating constraint-not-found error

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96648)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @rytaft)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96648#-NNhCDJu0QJW87OnMjp3:-NNnGRCEEEoU369je2Sv:b-4wvash) ([raw file](https://github.com/cockroachdb/cockroach/blob/aed8976bff170c998f1e5824fc34333d52a7e61b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go#L44)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

So what happens if you do:

```sql
CREATE TABLE t (i INT PRIMARY KEY);
ALTER TABLE t ADD CONSTRAINT c CHECK (i > 0) NOT VALID;
```
and
```sql 
ALTER TABLE t DROP CONSTRAINT c, VALIDATE CONSTRAINT c;
```
Or:
```sql
ALTER TABLE t DROP CONSTRAINT c, ADD CONSTRAINT c CHECK (i > 1) NOT VALID, VALIDATE CONSTRAINT c;
```
Does the first case error and the second case validate the newly added check constraint?
</blockquote></details>

Yes, I have ensured that new schema changer handles both cases as expected.

I also tried to make legacy schema changer handles the first case correctly, but it's not easy to get the second case work properly (see https://github.com/cockroachdb/cockroach/issues/96831)


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -2903,3 +2903,40 @@ t_96115     t_96115_pkey     PRIMARY KEY      true
+ALTER TABLE t_96648 ADD CONSTRAINT check_i CHECK (i > 0), VALIDATE CONSTRAINT check_i;
+
+statement ok
+ALTER TABLE t_96648 ADD CONSTRAINT check_i1 CHECK (i > 0) NOT VALID, VALIDATE CONSTRAINT check_i1;
```
could you add some test that the validation fails after inserting some bad data with `i <= 0`?

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_validate_constraint.go:
```diff
@@ -0,0 +1,145 @@
+		b.Drop(spec.ckNotValidElem)
+		b.Add(&scpb.CheckConstraint{
+			TableID:               tableID,
+			ConstraintID:          nextConstraintID,
```
interesting solution here, so that means the constraint ID changes... I will need to handle the back-references when a UDF is used in a check constraint heh. Is it possible to use the same constraint ID here?

chengxiong-ruan replied:
oh, actually nvm. it's handled :)

chengxiong-ruan reviewed (commented):
Nice work. just one request on test and a question on constraint ID increments.

Xiang-Gu commented:
TFTR!

bors r+
